### PR TITLE
[STACK-2099][Victoria] Add missing APIs for A10 Provider Driver

### PR DESCRIPTION
**Pull request by ytsai-a10**
## Description
Add victoria missing APIs for A10 Provider driver

## Jira Ticket
**This is optional if you are an external contributor**
- Required: Add links to tickets closed here

## Technical Approach
Add victoria missing APIs for A10 Provider driver

per https://docs.openstack.org/octavia/latest/contributor/guides/providers.html

> VIP port creation
>
> Some provider drivers will want to create the Neutron port for the VIP, and others will want Octavia to create the port instead. In order to support both use cases, the create_vip_port() method will ask provider drivers to create a VIP port. If the driver expects Octavia to create the port, the driver will raise a NotImplementedError exception. Octavia will call this function before calling loadbalancer_create() in order to determine if it should create the VIP port. Octavia will call create_vip_port() with a loadbalancer ID and a partially defined VIP dictionary. Provider drivers that support port creation will create the port and return a fully populated VIP dictionary.

## Config Changes
<pre>
<b>N/A</b>
</pre>

## Test Cases
 - test `openstack loadbalancer provider capability list a10`
 - test `openstack loadbalancer create`

## Manual Testing
 - test `openstack loadbalancer provider capability list a10`
```
stack@ytsai-victoria:~/source/a10-octavia/victoria$ openstack loadbalancer provider capability list a10
+--------+---------------------------------+----------------------------------------------------------------------------------------------------+
| type   | name                            | description                                                                                        |
+--------+---------------------------------+----------------------------------------------------------------------------------------------------+
| flavor | virtual-server                  | Specify axapi that will apply to the slb virtual-server                                            |
| flavor | virtual-server.name-expressions | Specify name expression to match loadbalancers and axapi that will apply to the slb virtual-server |
| flavor | virtual-port                    | Specify axapi that will apply to the vport                                                         |
| flavor | virtual-port.name-expressions   | Specify name expression to match listeners and axapi that will apply to the vport                  |
| flavor | service-group                   | Specify axapi that will apply to the service-group                                                 |
| flavor | service-group.name-expressions  | Specify name expression to match pools and axapi that will apply to the service-group              |
| flavor | server                          | Specify axapi that will apply to the server                                                        |
| flavor | server.name-expressions         | Specify name expression to match members and axapi that will apply to the server                   |
| flavor | health-monitor                  | Specify axapi that will apply to the health monitor                                                |
| flavor | health-monitor.name-expressions | Specify name expression to match healthmonitor and axapi that will apply to the health monitor     |
| flavor | nat-pool                        | Specify axapi of default nat pool for loadbalancer                                                 |
| flavor | nat-pool-list                   | Specify axapi of nat pools for loadbalancer                                                        |
+--------+---------------------------------+----------------------------------------------------------------------------------------------------+
```

 - test `openstack loadbalancer create`
```
stack@ytsai-victoria:~/source/a10-octavia/victoria$ openstack loadbalancer create --vip-subnet-id e21b771e-d439-46b2-8fad-fd3c745367a3 --vip-address 192.168.91.59 --name vip4
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| availability_zone   | None                                 |
| created_at          | 2021-03-05T06:17:32                  |
| description         |                                      |
| flavor_id           | None                                 |
| id                  | 04e2bbac-0f8a-4482-9502-d4351b94426b |
| listeners           |                                      |
| name                | vip4                                 |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | 0cc3169a379647aea0ce0c10697b75b6     |
| provider            | a10                                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| vip_address         | 192.168.91.59                        |
| vip_network_id      | f661c00b-9440-49bd-b4f7-ac033ec8404a |
| vip_port_id         | 95a23aba-25a8-4cac-9d17-618287c6c2ef |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | e21b771e-d439-46b2-8fad-fd3c745367a3 |
+---------------------+--------------------------------------+
stack@ytsai-victoria:~/source/a10-octavia/victoria$ openstack loadbalancer show vip4
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| availability_zone   | None                                 |
| created_at          | 2021-03-05T06:17:32                  |
| description         |                                      |
| flavor_id           | None                                 |
| id                  | 04e2bbac-0f8a-4482-9502-d4351b94426b |
| listeners           |                                      |
| name                | vip4                                 |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | 0cc3169a379647aea0ce0c10697b75b6     |
| provider            | a10                                  |
| provisioning_status | ACTIVE                               |
| updated_at          | 2021-03-05T06:17:36                  |
| vip_address         | 192.168.91.59                        |
| vip_network_id      | f661c00b-9440-49bd-b4f7-ac033ec8404a |
| vip_port_id         | 95a23aba-25a8-4cac-9d17-618287c6c2ef |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | e21b771e-d439-46b2-8fad-fd3c745367a3 |
+---------------------+--------------------------------------+

```

### CODE REVIEW

File: a10_octavia/api/drivers/driver.py [hunk 1]
```diff
@@ -304,3 +318,37 @@ def validate_flavor(self, flavor_dict):
...
+        descriptions from the JSON validation schema and returns it as a
+        dictionary.
+
+        :return: Dictionary of availability zone metadata keys and descriptions
+        :raises DriverError: An unexpected error occurred.
+        """
+        props = {}
+        return props
+
+    def validate_availability_zone(self, availability_zone_dict):
```
Comment (rahulkatre-a10) [inline a10_octavia/api/drivers/driver.py:336]: we could parent's class over here 
super().validate_availability_zone(availability_zone_dict)

File: a10_octavia/api/drivers/driver.py [hunk 2]
```diff
@@ -149,6 +157,12 @@ def member_update(self, old_member, new_member):
                    constants.MEMBER_UPDATES: member_dict}
         self.client.cast({}, 'update_member', **payload)
 
+    def member_batch_update(self, pool_id, members):
```
Comment (rahulkatre-a10) [inline a10_octavia/api/drivers/driver.py:160]: we can use
super().member_batch_update(pool_id, members)

File: a10_octavia/api/drivers/driver.py [hunk 3]
```diff
@@ -304,3 +319,37 @@ def validate_flavor(self, flavor_dict):
                                   'due to: {}'.format(str(e)),
                 operator_fault_string='Failed to validate the flavor metadata '
                                       'due to: {}'.format(str(e)))
+
+    # Availability Zone
+    def get_supported_availability_zone_metadata(self):
```
Comment (hthompson-a10) [inline a10_octavia/api/drivers/driver.py:324]: We need to raise a NotImplemented() here. The current information would imply to the operator that they have their availability zones misconfigured. Not that our driver doesn't support this operation.

**Review by hthompson-a10 — CHANGES_REQUESTED** [review]
I agree with Rahul. We need to be consistent in our error reporting. That said, we don't need to use the `super` methods unless we want to keep these functions around with a TODO note. In fact, this driver should return "NotImplemented" for these even without this PR.

That said, I recommend we implement our own fault strings just for increase clarity.

```user_fault_string='The a10 provider does not support <operation> <object> ```
```operator_fault_string='The a10 provider does not support <operation> <object>```

Something import is that we refer to the `a10` provider as that is our driver's exact alias when you print out the provider list.

**Comment by ytsai-a10** [comment]
Agree, the behavior should control by our code not up stream library.

> I agree with Rahul. We need to be consistent in our error reporting. That said, we don't need to use the `super` methods unless we want to keep these functions around with a TODO note. In fact, this driver should return "NotImplemented" for these even without this PR.
> 
> That said, I recommend we implement our own fault strings just for increased clarity.
> 
> `user_fault_string='The a10 provider does not support <operation> <object>'`
> `operator_fault_string='The a10 provider does not support <operation> <object>'`
> 
> Something import is that we refer to the `a10` provider as that is our driver's exact alias when you print out the provider list.

File: a10_octavia/api/drivers/driver.py (hunk 1 above)
Comment (ytsai-a10) [reply a10_octavia/api/drivers/driver.py:341]: - I prefer the behavior control by our-own not up stream library
- To have our-own error message

File: a10_octavia/api/drivers/driver.py (hunk 2 above)
Comment (ytsai-a10) [reply a10_octavia/api/drivers/driver.py:164]: - I prefer the behavior control by our-own not up stream library
- To have our-own error message

File: a10_octavia/api/drivers/driver.py (hunk 3 above)
Comment (ytsai-a10) [reply a10_octavia/api/drivers/driver.py:327]: Per https://docs.openstack.org/octavia/latest/contributor/guides/providers.html
> **get_supported_availability_zone_metadata¶**
> Retrieves a dictionary of supported availability zone keys and their description.

So, I think we should return empty dictionary for we support nothing. And if we raise exception command ```openstack loadbalancer provider capability list a10``` will failed with 500 Server Internal Error.

**Review by rahulkatre-a10 — APPROVED** [review]

**Pull request merged**